### PR TITLE
HTTP/2 limit header accumulated size

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
The Http2FrameListener requires that the Http2FrameReader accumulate ByteBuf objects. There is currently no way to limit the amount of bytes that is accumulated.

Modifications:
- DefaultHttp2FrameReader supports maxHeaderSize which will fail fast as soon as the maximum size is exceeded.
- DefaultHttp2HeadersDecoder will respect the return value of endHeaderBlock() and fail if the max size is exceeded.

Result:
Frames which carry header data now respect a maximum number of bytes that can be accumulated.

### CODE REVIEW

**Comment by Scottmitch** [comment]
@nmittler - PTAL

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java [hunk 1]
```diff
@@ -421,9 +422,8 @@ public void processFragment(boolean endOfHeaders, ByteBuf fragment,
                     final HeadersBlockBuilder hdrBlockBuilder = headersBlockBuilder();
                     hdrBlockBuilder.addFragment(fragment, ctx.alloc(), endOfHeaders);
                     if (endOfHeaders) {
-                        listener.onHeadersRead(ctx, headersStreamId, hdrBlockBuilder.headers(),
-                                        streamDependency, weight, exclusive, padding, headersFlags.endOfStream());
-                        close();
+                        listener.onHeadersRead(ctx, headersStreamId, hdrBlockBuilder.headers(), streamDependency,
+                                weight, exclusive, padding, headersFlags.endOfStream());
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java]: I'm assuming you removed the call to `close()` because it's automatically done by `headers()`?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java [hunk 2]
```diff
@@ -635,8 +642,11 @@ final void close() {
          *            This is used for an optimization for when the first fragment is the full
          *            block. In that case, the buffer is used directly without copying.
          */
-        final void addFragment(ByteBuf fragment, ByteBufAllocator alloc, boolean endOfHeaders) {
+        final void addFragment(ByteBuf fragment, ByteBufAllocator alloc, boolean endOfHeaders) throws Http2Exception {
             if (headerBlock == null) {
+                if (fragment.readableBytes() > headersDecoder.configuration().maxHeaderSize()) {
+                    headerSizeExceeded();
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java]: Just to be consistent with the other checks, maybe make a method `verifyHeaderSize(ByteBuf fragment)` that performs the check and then throws if the new fragment violates the threshold?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java [hunk 3]
```diff
@@ -627,6 +624,16 @@ final void close() {
         private ByteBuf headerBlock;
 
         /**
+         * The local header size maximum has been exceeded while accumulating bytes.
+         * @throws Http2Exception A connection error indicating too much data has been received.
+         */
+        private void headerSizeExceeded() throws Http2Exception {
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java]: Move this down with the other `verifyXXX` methods?

**Comment by nmittler** [comment]
@Scottmitch looks good ... just a few comments.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java [hunk 4]
```diff
@@ -50,10 +56,23 @@ public Http2HeaderTable headerTable() {
...
+    @Override
     public Configuration configuration() {
         return this;
     }
 
+    /**
+     * Respond to headers block resulting in the maximum header size being exceeded.
+     * @throws Http2Exception If we can not recover from the truncation.
+     */
+    protected void maxHeaderSizeExceeded() throws Http2Exception {
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java]: final ?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java (hunk 1 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java]: Yip. No need to do it more than 1 time.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java (hunk 3 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java]: Those methods are outside the scope of `HeadersBlockBuilder` and I would have to pass an instance in as a parameter to be able to do the `close()` method.  This method is not doing any `verify`...the check has been done outside the scope and is triggered in different conditions depending on the state of the `HeadersBlockBuilder`.  I could move...but not sure if it provides any benefit. WDTY?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java (hunk 2 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java]: I would have to recheck the `null` condition, and don't anticipate this new method to be used outside the `addFragment` method scope.  I'm tempted to leave it as is...WDYT?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java (hunk 4 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java]: The intention was to make it extensible.  The default behavior is to `throw`, but some implementations may want to carry on with what they have?

Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java]: ok if it is valid to do so its fine :)

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java (hunk 3 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java]: ah right ... sounds good.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java (hunk 2 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java]: Yup ... fine as-is.

**Comment by nmittler** [comment]
LGTM

**Comment by Scottmitch** [comment]
Cherry-picked 4.1 (ecacd11) master (a9e57f1).

Thanks for review!
